components/service/glean/src/main/java/mozilla/components/service/glean/CounterMetricType.kt

```diff
@@ -55,4 +55,14 @@ data class CounterMetricType(
             )
         }
     }
+
+    override fun getWithName(newName: String): CounterMetricType {
+        return CounterMetricType(
```

Comment:
Sure, I'll play with this some more.  Couldn't find another way in our "data-classes-that-inherit-from-an-interface" framing we currently have.

Reply:
Sure! Even if he'll hate me a bit, I'd feel more comfortable with this if @pocmo (or anyone else with Kotlin experience) would give at least the `LabeledMetricType.kt` file an in-depth review, so maybe worth asking him too!

Reply:
The gotcha here is that generics in Kotlin require a common interface, since it doesn't generate a new function for each type T.  Since data classes don't inherit from a common interface or base class, we can't call `copy` on them in this context.

Reply:
You can do this:

```kotlin
    private fun getMetric(name: String): T {
        return when (subMetric) {
            is BooleanMetricType -> subMetric.copy(name = name) as T
            is CounterMetricType -> subMetric.copy(name = name) as T
            is StringListMetricType -> subMetric.copy(name = name) as T
            is StringMetricType -> subMetric.copy(name = name) as T
            is TimespanMetricType -> subMetric.copy(name = name) as T
            is UuidMetricType -> subMetric.copy(name = name) as T
            else -> throw IllegalStateException()
        }
    }
```

which is a lot less verbose and puts all of the weirdness in one place, and removes the need for the `getWithName` method and the template parameter on `CommonMetricData`, but it's probably not quite as efficient (though I'm only guessing).